server/pairing/payload_receiver.go

```diff
@@ -0,0 +1,383 @@
+		zap.String("accountPayloadMarshaller.accountPayloadMarshaller.password", apr.accountPayloadMarshaller.password),
+		zap.Binary("accountPayloadMarshaller.Received()", apr.Received()),
+	)
+
+	signal.SendLocalPairingEvent(Event{Type: EventReceivedAccount, Action: ActionPairingAccount, Data: apr.accountPayload.multiaccount})
+
+	return apr.accountStorer.StoreToSource()
+}
+
+func (apr *AccountPayloadReceiver) Received() []byte {
+	return apr.encryptor.getDecrypted()
+}
+
+func (apr *AccountPayloadReceiver) LockPayload() {
+	apr.encryptor.lockPayload()
+}
+
+// AccountPayloadStorer is responsible for parsing, validating and storing AccountPayload data
+type AccountPayloadStorer struct {
+	*AccountPayload
+	multiaccountsDB *multiaccounts.Database
+
+	keystorePath   string
+	kdfIterations  int
+	loggedInKeyUID string
+}
+
+func NewAccountPayloadStorer(p *AccountPayload, config *ReceiverConfig) (*AccountPayloadStorer, error) {
+	ppr := &AccountPayloadStorer{
+		AccountPayload: p,
+	}
+
+	if config == nil {
+		return ppr, nil
+	}
+
+	ppr.multiaccountsDB = config.DB
+	ppr.kdfIterations = config.KDFIterations
+	ppr.keystorePath = config.KeystorePath
+	ppr.loggedInKeyUID = config.LoggedInKeyUID
+	return ppr, nil
+}
+
+func (aps *AccountPayloadStorer) StoreToSource() error {
+	keyUID := aps.multiaccount.KeyUID
+	if aps.loggedInKeyUID != "" && aps.loggedInKeyUID != keyUID {
+		return ErrLoggedInKeyUIDConflict
+	}
+	if aps.loggedInKeyUID == keyUID {
+		// skip storing keys if user is logged in with the same key
+		return nil
+	}
+
+	err := validateKeys(aps.keys, aps.password)
+	if err != nil {
+		return err
+	}
+
+	if err = aps.storeKeys(aps.keystorePath); err != nil && err != ErrKeyFileAlreadyExists {
+		return err
+	}
+
+	// skip storing multiaccount if key already exists
+	if err == ErrKeyFileAlreadyExists {
+		aps.exist = true
+		aps.multiaccount, err = aps.multiaccountsDB.GetAccount(keyUID)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return aps.storeMultiAccount()
+}
+
+func (aps *AccountPayloadStorer) storeKeys(keyStorePath string) error {
+	if keyStorePath == "" {
+		return fmt.Errorf("keyStorePath can not be empty")
+	}
+
+	_, lastDir := filepath.Split(keyStorePath)
+
+	// If lastDir == "keystore" we presume we need to create the rest of the keystore path
+	// else we presume the provided keystore is valid
+	if lastDir == "keystore" {
+		if aps.multiaccount == nil || aps.multiaccount.KeyUID == "" {
+			return fmt.Errorf("no known Key UID")
+		}
+		keyStorePath = filepath.Join(keyStorePath, aps.multiaccount.KeyUID)
+		_, err := os.Stat(keyStorePath)
+		if os.IsNotExist(err) {
+			err := os.MkdirAll(keyStorePath, 0777)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		} else {
+			return ErrKeyFileAlreadyExists
+		}
+	}
+
+	for name, data := range aps.keys {
+		accountKey := new(keystore.EncryptedKeyJSONV3)
+		if err := json.Unmarshal(data, &accountKey); err != nil {
```

Comment:
Yes. We can do file validation first then data store in a second loop.

If one of the keys fails, in either case, something quite bad has happened. So yes I agree the cleanest solution would be to attempt to revert the process. Depending on the error the application may no longer have io access, but that's probably why you've written "*attempt*". IF the clean up fails we just need to abort the process and start cascading errors.

Reply:
Yes, I agree with you it's definitely an edge case.

I was also thinking if the whole pairing process isn't idempotent (it is right?), so the user can just initiate the process again and the files would be recreated/replaced in case an exceptional error happened. As long as the user get the proper feedback from clients that could be more than good enough.

>  IF the clean up fails we just need to abort the process and start cascading errors.

Exactly @Samyoul, we can just try to revert on a best effort basis.